python/ray/experimental/iter.py

```diff
@@ -0,0 +1,556 @@
+
+    The objects will be divided round-robin among the number of shards.
+
+    Arguments:
+        items (list): The list of items to iterate over.
+        num_shards (int): The number of worker actors to create.
+    """
+    shards = [[] for _ in range(num_shards)]
+    for i, item in enumerate(items):
+        shards[i % num_shards].append(item)
+    name = "from_items[{}, {}, shards={}]".format(
+        items and type(items[0]).__name__ or "None", len(items), num_shards)
+    return from_iterators(shards, name=name)
+
+
+def from_range(n: int, num_shards: int = 2) -> "ParIterator[int]":
+    """Create a parallel iterator over the range 0..n.
+
+    The range will be partitioned sequentially among the number of shards.
+
+    Arguments:
+        n (int): The max end of the range of numbers.
+        num_shards (int): The number of worker actors to create.
+    """
+    generators = []
+    for i in range(num_shards):
+        start = i * (n // num_shards)
+        if i == num_shards - 1:
+            end = n
+        else:
+            end = (i + 1) * (n // num_shards)
+        generators.append(range(start, end))
+    name = "from_range[{}, shards={}]".format(n, num_shards)
+    return from_iterators(generators, name=name)
+
+
+def from_iterators(generators: List[Iterable[T]],
+                   name=None) -> "ParIterator[T]":
+    """Create a parallel iterator from a set of iterators.
+
+    An actor will be created for each iterator.
+
+    Arguments:
+        generators (list): A list of Python generator objects or lambda
+            functions that produced a generator when called.
+        name (str): Optional name to give the iterator.
+    """
+    worker_cls = ray.remote(_ParIteratorWorker)
+    actors = [worker_cls.remote(g) for g in generators]
+    if not name:
+        name = "from_iterators[shards={}]".format(len(generators))
+    return from_actors(actors, name=name)
+
+
+def from_actors(actors: List["ray.actor.ActorHandle"],
+                name=None) -> "ParIterator[T]":
+    """Create a parallel iterator from an existing set of actors.
+
+    Each actor must implement the par_iter_init() and par_iter_next() methods
+    from the _ParIteratorWorker interface.
+
+    Arguments:
+        actors (list): List of actors that each implement _ParIteratorWorker.
+        name (str): Optional name to give the iterator.
+    """
+    if not name:
+        name = "from_actors[shards={}]".format(len(actors))
+    return ParIterator([_ActorSet(actors, [])], name)
+
+
+class ParIterator(Generic[T]):
+    """A parallel iterator over a set of remote actors.
+
+    This can be used to iterate over a fixed set of task results
+    (like an actor pool), or a stream of data (e.g., a fixed range of numbers,
+    an infinite stream of RLlib rollout results).
+
+    This class is **serializable** and can be passed to other remote
+    tasks and actors. However, each shard should be read from at most one
```

Comment:
What's the intended way to handle this? Pass a "shard index" to each actor that tells it which to read?

Reply:
Might be nicer to support passing individual shards (if it isn't already) so the receiving end can just treat it as a normal iterator

Reply:
This is just `ParIterator.get_shard(5)` right?

Reply:
Yeah we support both!